// catch-express-error/source/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.handleGlobalErrors = exports.catchAsync = exports.AppError = void 0;
class AppError {
    message;
    statusCode;
    status;
    isOperational;
    errName;
    code;
    name;
    details;
    constructor(message, statusCode, details, name, code) {
        this.message = message;
        this.statusCode = statusCode ?? 400;
        this.status = `${statusCode}`.startsWith("4") ? "fail" : "error";
        this.isOperational = true;
        this.name = name ?? "App Error";
        this.code = code;
        this.details = details;
        Error.captureStackTrace(this, this.constructor);
    }
}
exports.AppError = AppError;
const catchAsync = (fn) => {
    return (req, res, next) => {
        fn(req, res, next).catch(next);
    };
};
exports.catchAsync = catchAsync;
const showErrMsg = (err, logger) => {
    const errMsg = `💥 ${err.stack ?? (typeof err === "string" ? err : JSON.stringify(err))}`;
    logger ? logger.log("error", err) : console.error(errMsg);
};
const handleGlobalErrors = (logger, isProduction = false) => {
    return (err, __, res, _) => {
        err.statusCode = err.statusCode || 500;
        err.status = err.status || "error";
        showErrMsg(err, logger);
        if (!isProduction)
            return res.status(err.statusCode).json({
                error: err,
                stack: err.stack,
            });
        if (err.isOperational) {
            return res.status(err.statusCode).json({
                status: err.status,
                message: err.message,
                statusCode: err.statusCode,
            });
        }
        return res.status(500).json({
            status: "error",
            message: "Something went very wrong!",
            statusCode: 500,
        });
    };
};
exports.handleGlobalErrors = handleGlobalErrors;
